# Assignment 1 part B

In [1]:
import numpy as np

## B.1. Vectorization in Numpy

Vectorization refers to the practice of replacing explicit loops with array expressions, which leads to more efficient, readable, and concise code. 

Let's see why this is useful. We will compare adding up two arrays in the two approaches.

In [3]:
import numpy as np
# Set random seed. This ensures that we always get the same random numbers when we run this cell in the notebook.
np.random.seed(123)
# Create Arrays
a = np.random.rand(1000000)
b = np.random.rand(1000000)

**Question (a):** 

Write a function `add_with_loop`` that takes two arrays and returns a new array where each element is the sum of the elements at the corresponding position in the original arrays. Use Python loops to iterate over the arrays.

Write a function `add_with_vectorization`` that performs the same operation, but uses NumPy's vectorized operations.

In [36]:
def add_with_loop(a, b):
    """
    Perform elementwise addition of two one-dimensional arrays.

    Parameters:
    - a (np.ndarray): The first one-dimensional array.
    - b (np.ndarray): The second one-dimensional array.

    Returns:
    - result (np.ndarray): An array containing the elementwise sum of a and b.

    Note:
    1. Both arrays 'a' and 'b' must have the same length for valid elementwise addition.
    2. The result array is calculated as result[i] = a[i] + b[i] for each element i
       in the input arrays.
    """
    assert isinstance(a, np.ndarray) and isinstance(b, np.ndarray), "a and/or b incorrect data type"
    assert a.ndim == 1 and b.ndim == 1, "a and/or b are not one-dimensional arrays"
    assert len(a) == len(b), "a and b are not the same length"
    result = np.empty(len(a))
    for i in range(len(a)):
        result[i] = a[i] + b[i]
    return result

# Task 2.2: Using Vectorization
def add_with_vectorization(a, b):
    """
    Perform elementwise addition of two arrays using NumPy's add function.

    Parameters:
    - a (np.ndarray): The first array.
    - b (np.ndarray): The second array.

    Returns:
    - result (np.ndarray): An array containing the elementwise sum of a and b.

    Note:
    1. Both arrays 'a' and 'b' must have the same shape for valid elementwise addition.
    2. The result array is obtained by adding corresponding elements from 'a' and 'b'
       using NumPy's add function.
    """
    return np.add(a, b)

**Question (b):** 

Measure Execution Time: Use the `time` module to measure the execution time of both functions. Compare the time taken by add_with_loop and add_with_vectorization for the same input arrays.

Write a brief analysis of the performance difference. Discuss why vectorization leads to such a difference in execution time.

The following code demonstrates the time syntax:

In [6]:
import time
start_time = time.time()
add_with_loop(a, b)
time_with_loop = time.time() - start_time
print(f"Time with loop: {time_with_loop} seconds")

Time with loop: 0.18627619743347168 seconds


However, for full credit, you should not rely on a single iteration. Why? We discuss this briefly in class.

In [7]:
nit = 100

# Measure execution time for loop-based addition
start_time = time.time()
for i in range(nit):
    add_with_loop(a, b)
time_with_loop = time.time() - start_time

# Measure execution time for vectorized addition
start_time = time.time()
for i in range(nit):
    add_with_vectorization(a, b)
time_with_vectorization = time.time() - start_time

# Print the results
print(f"Time with loop: {time_with_loop} seconds")
print(f"Time with vectorization: {time_with_vectorization} seconds")
print(f"Ratio: {time_with_loop/time_with_vectorization}")

Time with loop: 14.140784740447998 seconds
Time with vectorization: 0.07491207122802734 seconds
Ratio: 188.7651016537027


Discussion of results: The time with vectorization is much quicker and more efficient than the time with the loop, being almost 200 times faster over 100 iterations. As explained in the Python Data Science Handbook, this is because of the type-checking (examining the object's type) and function dispatching (looking up the correct function to use for that type) that CPython does at each cycle of the loop. Meanwhile, the vectorized approach pushes the loop into the compiled layer that underlies NumPy, where the type specification is known before the code executes, making computation much faster.

**Question (c):** 

The "magic function" `%%timeit` is another way to time your functions. You can read about it in the [Python Data Science Handbook/ Jake VanderPlas](https://jakevdp.github.io/PythonDataScienceHandbook/) (interactive features). Compare the functions again using timeit.

In [12]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit add_with_loop(a, b)

137 ms ± 1.07 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [13]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit add_with_vectorization(a, b)

694 µs ± 9.58 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


## B.2. Matrix Multiplication

In [44]:
np.random.seed(123)
# Create Arrays
n1=10
n2=5
n3=15
A = np.random.rand(n1,n2)
B = np.random.rand(n2,n3)

**Question :** (a) Explicit matrix multiplication



In [45]:
def explicit_matrix_multiplication(A, B):
    """
    Multiply matrices A and B together explicitly.
    
    Parameters:
    - A (np.ndarray): The first matrix with shape (m, n).
    - B (np.ndarray): The second matrix with shape (n, p).

    Returns:
    - w (np.ndarray): The result of the matrix multiplication A * B,
      with shape (m, p).

    Note:
    1. The number of columns in matrix A must be equal to the number of
       rows in matrix B for valid matrix multiplication.
    2. The resulting matrix w is calculated as w[i, j] = sum(A[i, k] * B[k, j])
       for each element in the result matrix, where the sum is taken over the
       common dimension 'n'.
    """
    assert isinstance(A, np.ndarray) and isinstance(B, np.ndarray), "A and/or B incorrect data type"
    m, n = A.shape
    o, p = B.shape
    assert n == o, "The number of columns in matrix A must be equal to the number of rows in matrix B."
    w = np.zeros((m, p))
    for i in range(m):
        for j in range(p):
            mysum = 0
            for k in range(n):
                mysum = mysum+A[i,k]*B[k,j]
            w[i, j]=mysum
    return w

**Question :** (b) compare your result to numpy multiplication. Are they exactly the same? What do you think is going on?

In [46]:
C = explicit_matrix_multiplication(A, B)
D = A@B
print(C==D)
print(np.abs(C - D) < 10**(-12)*1)

[[False  True  True False  True  True  True  True False  True  True False
   True False False]
 [ True  True  True False  True  True  True  True  True False  True False
   True  True  True]
 [ True  True False  True  True False  True  True  True  True False  True
   True  True  True]
 [ True  True False  True  True False  True False  True  True  True  True
  False  True False]
 [False  True  True  True  True  True  True  True  True False False  True
  False  True  True]
 [ True False  True False False  True  True  True  True  True  True False
  False  True  True]
 [ True  True False  True  True  True False  True  True False  True False
   True  True  True]
 [ True  True  True  True False  True False False  True False  True  True
   True  True False]
 [ True  True  True  True  True  True  True  True  True  True False False
   True  True  True]
 [ True False  True False  True  True False  True  True False  True  True
   True  True False]]
[[ True  True  True  True  True  True  True  True

Discussion of Results: While the two methods appear to be the same, they are not exactly the same when using == comparison, but are "close" with a difference less than 10^(-12). This is because at some point the floating point is cut off before computation, leading to results that are not 100% accurate or the same based on where the loop versus numpy chooses to cut off the float (as discussed in class). This difference in computation could be because of the Python vs. compiler difference in both methods.

**Question :** explicit-matrix-part_c

Compare the timing to the built-in matrix multiplication in numpy.

In [34]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit explicit_matrix_multiplication(A, B)

151 µs ± 7.03 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [35]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit A@B

794 ns ± 9.36 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Discussion of Results: The built-in matrix multiplication in numpy is much faster (794 ns +/- 9.36 ns per loop) compared to the explicit matrix multiplication (151 us +/- 7.03 per loop).

## B.3. Calculating Distances: Broadcasting in Numpy

In the next cell we create a dataset with two sets of two dimensional points. Each row represents a point.

The Euclidean distance between two points p and q (each a vector of length 2) is $\sqrt{(p-q,q-q)} = \sqrt{\sum_{j=1}^2 (p[j]-q[j])^2 }$.

In [12]:
np.random.seed(123)
# Create Arrays
distn1=1000
distn2=1500
distdim = 2
a = np.random.rand(distn1,distdim)
b = np.random.rand(distn2,distdim)

**Question :** (a) Distance between two points.

Write a function that takes two points and computes the distance between them. For full credit, use numpy functionlity, not the explicit loop over axes, and make it work for any dimension.

In [13]:
def dist_two_points(p1 , p2):
    """
    Calculate the Euclidean distance between two points in N-dimensional space.

    Parameters:
    - p1 (np.ndarray): The coordinates of the first point.
    - p2 (np.ndarray): The coordinates of the second point.

    Returns:
    - distance (float): The Euclidean distance between the two points.

    Note:
    1. Both points 'p1' and 'p2' have the same dimensionality (number of
       coordinates) for valid distance calculation.
    2. The Euclidean distance is calculated as the square root of the sum of squared
       differences between corresponding coordinates of the two points.
    """
    assert isinstance(p1, np.ndarray) and isinstance(p2, np.ndarray), "p1 and/or p2 incorrect data type"
    assert p1.ndim == p2.ndim, "p1 and p2 are not of the same dimension"
    if p1.ndim == 1:
        assert len(p1) == len(p2), "p1 and p2 are not of the same dimension"
    diff = np.subtract(p1, p2)
    diff2 = np.power(diff, 2)
    sumdiff = np.sum(diff2, axis=((p1.ndim)-1))
    return np.power(sumdiff, 0.5)

**Question :** (a) Write a function that returns a matrix of distances between two sets of points X and Y, such that the element in the ith row and jth column is the distance between the ith row of X and the jth row of Y.

Time your function.

In [16]:
def dist_matrix_loop(X,Y):
    """
    Calculate the pairwise Euclidean distance matrix between two sets of points using loops.

    Parameters:
    - X (np.ndarray): The first set of points with shape (m, d) where m is the number
      of points and d is the dimensionality of each point.
    - Y (np.ndarray): The second set of points with shape (n, d) where n is the number
      of points and d is the dimensionality of each point.

    Returns:
    - result (np.ndarray): A matrix containing the Euclidean distances between each
      pair of points from sets X and Y, with shape (m, n).

    Note:
    1. Both sets of points 'X' and 'Y' must have the same dimensionality (number of
       coordinates) for valid distance calculation.
    2. The Euclidean distance between each pair of points (X[i], Y[j]) is calculated
       using the `dist_two_points` function.
    """
    assert isinstance(X, np.ndarray) and isinstance(Y, np.ndarray), "X and/or Y incorrect data type"
    assert X.ndim == X.ndim and len(X[0]) == len(Y[0]), "X and Y are not of the same dimension"
    m = len(X)
    n = len(Y)
    result = np.zeros((m, n))
    for i in range(m):
        for j in range(n):
            dist = dist_two_points(X[i], Y[j])
            result[i, j] = dist
    assert result[0,0] == dist_two_points(X[0], Y[0])
    return result

In [16]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit dist_matrix_loop(a,b)

4.98 s ± 30.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


**Question :** (c) Write a vectorized version, using broadcasting to compute the matrix of distances. Time your solution.

In [23]:
def dist_matrix_broadcast(X,Y):
    """
    Calculate the pairwise Euclidean distance matrix between two sets of points using broadcasting.

    Parameters:
    - X (np.ndarray): The first set of points with shape (m, d) where m is the number
      of points and d is the dimensionality of each point.
    - Y (np.ndarray): The second set of points with shape (n, d) where n is the number
      of points and d is the dimensionality of each point.

    Returns:
    - result (np.ndarray): A matrix containing the Euclidean distances between each
      pair of points from sets X and Y, with shape (m, n).

    Note:
    1. Both sets of points 'X' and 'Y' must have the same dimensionality (number of
       coordinates) for valid distance calculation.
    2. The Euclidean distance between each pair of points (X[i], Y[j]) is calculated
       using the `dist_two_points` function and broadcasting.
    """
    assert X.ndim == X.ndim and len(X[0]) == len(Y[0]), "X and Y are not of the same dimension"
    X1 = X[:, None]
    Y1 = Y[None, :]
    result = dist_two_points(X1, Y1)
    return result

In [ ]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit dist_matrix_broadcast(a,b)